pkg/apis/events/validation/validation.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/apis/core"
+	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
+)
+
+const (
+	reportingInstanceLengthLimit = 128
+	actionLengthLimit            = 128
+	reasonLengthLimit            = 128
+	noteLengthLimit              = 1024
+)
+
+// ValidateEvent validates new created events.
+func ValidateEvent(event *core.Event) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&event.ObjectMeta, true, apimachineryvalidation.NameIsDNSSubdomain, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateEventSpec(event)...)
+	return allErrs
+}
+
+// ValidateEventUpdate validates updated events.
+func ValidateEventUpdate(newEvent, oldEvent *core.Event) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMetaUpdate(&newEvent.ObjectMeta, &oldEvent.ObjectMeta, field.NewPath("metadata"))
+
+	restoreSeries := newEvent.Series
+	newEvent.Series = oldEvent.Series
+	if !apiequality.Semantic.DeepEqual(newEvent, oldEvent) {
```

Comment:
I'm not a fan of this approach because:
1. if we decide to add new mutable field in the future we will have to change this
2. (probably more important) the error is not super informative.

I would much rather to go over all fields - check their equality and report accurate error for them, i.e.:
```
if !newEvent.EventTime.Equal(oldEvent.EventTime) {
  allErrors = append(allErros, field.Forbidden(field.NewPath("eventTime"), ...)
}
if newEvent.ReportingController != oldEvent.ReportingController {
  allErrors = append(allErrrs, field.Forbidden(field.NewPath("reportingController"), ...)
}
...
```

Reply:
Done.